tests/functional/test_cache.py

```diff
@@ -0,0 +1,217 @@
+import os
+import shutil
+from glob import glob
+
+import pytest
+
+
+@pytest.fixture
+def cache_dir(script):
+    result = script.run(
+        'python', '-c',
+        'from pip._internal.locations import USER_CACHE_DIR;'
+        'print(USER_CACHE_DIR)'
+    )
+    return result.stdout.strip()
+
+
+@pytest.fixture
+def wheel_cache_dir(cache_dir):
+    return os.path.normcase(os.path.join(cache_dir, 'wheels'))
+
+
+@pytest.fixture
+def wheel_cache_files(wheel_cache_dir):
+    destination = os.path.join(wheel_cache_dir, 'arbitrary', 'pathname')
+
+    if not os.path.exists(destination):
+        return []
+
+    filenames = glob(os.path.join(destination, '*.whl'))
+    files = []
+    for filename in filenames:
+        files.append(os.path.join(destination, filename))
+    return files
+
+
+@pytest.fixture
+def populate_wheel_cache(wheel_cache_dir):
+    destination = os.path.join(wheel_cache_dir, 'arbitrary', 'pathname')
+    os.makedirs(destination)
+
+    files = [
+        ('yyy-1.2.3', os.path.join(destination, 'yyy-1.2.3-py3-none-any.whl')),
+        ('zzz-4.5.6', os.path.join(destination, 'zzz-4.5.6-py3-none-any.whl')),
+        ('zzz-4.5.7', os.path.join(destination, 'zzz-4.5.7-py3-none-any.whl')),
+        ('zzz-7.8.9', os.path.join(destination, 'zzz-7.8.9-py3-none-any.whl')),
+    ]
+
+    for _name, filename in files:
+        with open(filename, 'w'):
+            pass
+
+    return files
+
+
+@pytest.fixture
+def empty_wheel_cache(wheel_cache_dir):
+    if os.path.exists(wheel_cache_dir):
+        shutil.rmtree(wheel_cache_dir)
+
+
+def list_matches_wheel(wheel_name, result):
+    """Returns True if any line in `result`, which should be the output of
+    a `pip cache list` call, matches `wheel_name`.
+
+    E.g., If wheel_name is `foo-1.2.3` it searches for a line starting with
+          `- foo-1.2.3-py3-none-any.whl `."""
+    lines = result.stdout.splitlines()
+    expected = ' - {}-py3-none-any.whl '.format(wheel_name)
+    return any(map(lambda l: l.startswith(expected), lines))
+
+
+@pytest.fixture
+def remove_matches_wheel(wheel_cache_dir):
+    """Returns True if any line in `result`, which should be the output of
+    a `pip cache remove`/`pip cache purge` call, matches `wheel_name`.
+
+    E.g., If wheel_name is `foo-1.2.3`, it searches for a line equal to
+    `Removed <wheel cache dir>/arbitrary/pathname/foo-1.2.3-py3-none-any.whl`.
+    """
+
+    def _remove_matches_wheel(wheel_name, result):
+        lines = result.stdout.splitlines()
+
+        wheel_filename = '{}-py3-none-any.whl'.format(wheel_name)
+
+        # The "/arbitrary/pathname/" bit is an implementation detail of how
+        # the `populate_wheel_cache` fixture is implemented.
+        expected = 'Removed {}/arbitrary/pathname/{}'.format(
```

Comment:
The `/` characters should probably be `os.sep` - on Windows, `\` is the path separator.

Reply:
Okay, now I feel stupid for @'ing you before taking a deeper look myself at the error.

Reply:
np, easy to miss :-) And I like @duckinator's fix of using `os.path.join` better than my suggestion anyway

Reply:
Looks like that was it! Switched to `os.path.join()` instead of hard-coding the separator, which I suspect is functionally-equivalent, and it worked fine.